js_modules/dagster-ui/packages/ui-core/src/assets/lineage/useColumnLineageDataForAssets.tsx

```diff
@@ -0,0 +1,153 @@
+
+import {
+  AssetColumnLineageQuery,
+  AssetColumnLineageQueryVariables,
+} from './types/useColumnLineageDataForAssets.types';
+import {toGraphId} from '../../asset-graph/Utils';
+import {AssetKeyInput, TableSchemaMetadataEntry} from '../../graphql/types';
+import {
+  isCanonicalColumnLineageEntry,
+  isCanonicalColumnSchemaEntry,
+} from '../../metadata/TableSchema';
+
+export type AssetColumnLineageServer = {
+  [column: string]: {
+    // Note: This is [["key_part_1", "key_part_2"]] but the outer array
+    // only contains one item, it's a serialization odditiy.
+    upstream_asset_key: string[][];
+    upstream_column_name: string;
+  }[];
+};
+
+export type AssetColumnLineageLocal = {
+  [column: string]: {
+    name: string;
+    type: string | null;
+    description: string | null;
+    upstream: {
+      assetKey: AssetKeyInput;
+      columnName: string;
+    }[];
+  };
+};
+
+export type AssetColumnLineages = {[graphId: string]: AssetColumnLineageLocal | undefined};
+
+/**
+ * The column definitions and the column lineage are in two separate metadata entries,
+ * and the definitions may be specified in definition-time or materialization-time metadata.
+ * Parse them both and combine the results into a single representation of asset columns
+ * that is easier for the rest of the front-end to use.
+ */
+const getColumnLineage = (
+  asset: AssetColumnLineageQuery['assetNodes'][0],
+): AssetColumnLineageLocal | undefined => {
+  const materialization = asset.assetMaterializations[0];
+  const lineageMetadata = materialization?.metadataEntries.find(isCanonicalColumnLineageEntry);
+  if (!lineageMetadata) {
+    return undefined;
+  }
+
+  const schemaMetadata: TableSchemaMetadataEntry | undefined =
+    materialization?.metadataEntries.find(isCanonicalColumnSchemaEntry) ||
+    asset.metadataEntries.find(isCanonicalColumnSchemaEntry);
+
+  const lineageParsed: AssetColumnLineageServer = JSON.parse(lineageMetadata.jsonString);
+  const schemaParsed = schemaMetadata
+    ? Object.fromEntries(schemaMetadata.schema.columns.map((col) => [col.name, col]))
+    : {};
+
+  return Object.fromEntries(
+    Object.entries(lineageParsed).map(([column, m]) => [
+      column,
+      {
+        name: column,
+        type: schemaParsed[column]?.type || null,
+        description: schemaParsed[column]?.description || null,
+        upstream: m.map((u) => ({
+          assetKey: {path: u.upstream_asset_key[0]!},
+          columnName: u.upstream_column_name,
+        })),
+      },
+    ]),
+  );
+};
+
+export function useColumnLineageDataForAssets(assetKeys: AssetKeyInput[]) {
+  const [loaded, setLoaded] = useState<AssetColumnLineages>({});
+  const missing = assetKeys.filter((a) => !loaded[toGraphId(a)]);
+  const missingJSON = JSON.stringify(missing);
+  const client = useApolloClient();
+  const fetching = useRef(false);
+
+  React.useEffect(() => {
+    const missingAssetKeys = JSON.parse(missingJSON);
```

Comment:
feels bad to have to parse this again to avoid the hook warning

Reply:
Ahh good call, I memoized a few things upstream and that allows us to just pass in the original `missing` array. Thanks!